src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+            {
+                Debug.Assert((_state & (int)State.Started) == 0 && _thread == null);
+
+                // Recursive ControlledExecution.Run calls are not supported
+                if (t_execution != null)
+                    throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        // As soon as the Started flag is set, this thread may be aborted asynchronously
+                        if (Interlocked.CompareExchange(ref _state, (int)State.Started, (int)State.None) == (int)State.None)
+                        {
+                            t_execution = this;
+                            _action();
+                        }
+                    }
+                    finally
+                    {
+                        if ((_state & (int)State.Started) != 0)
+                        {
+                            // Set the Finished flag to prevent a potential subsequent AbortThread call
+                            State oldState = (State)Interlocked.Or(ref _state, (int)State.Finished);
+
+                            if ((oldState & State.AbortRequested) != 0)
+                            {
+                                // Either in SFAR or SFA state
+                                while (true)
+                                {
+                                    // The enclosing finally may be cloned by the JIT for the non-exceptional code flow.
+                                    // In that case this code is not guarded against a thread abort, so make this FCall as
+                                    // soon as possible.
+                                    bool resetAbortRequest = ResetAbortThread();
+
+                                    // If there is an Abort in progress, we need to wait until it sets the TS_AbortRequested
+                                    // flag on this thread, then we can reset the flag and safely exit this frame.
+                                    if (((oldState & State.RunningAbort) == 0) || resetAbortRequest)
+                                    {
+                                        break;
+                                    }
+
+                                    // It should take very short time for AbortThread to set the TS_AbortRequested flag
+                                    Thread.Sleep(0);
+                                    oldState = (State)Volatile.Read(ref _state);
+                                }
+                            }
+
+                            t_execution = null;
+                        }
+                    }
+                }
+                catch (ThreadAbortException) when (_cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException(_cancellationToken);
+                }
```

Comment:
Maybe we should set the inner exception to the TAE?

Reply:
I'd prefer TAE instances never find their way to user code, even as inner exceptions.

Reply:
> that's the only thing that could trigger an abort

What's about a func-eval abort with `ControlledExecution.Run` on the stack?

Reply:
> What's about a func-eval abort with ControlledExecution.Run on the stack?

Does that trigger a catchable ThreadAbortException?  If so, then I understand the desire for the filter (a comment is worthwhile), though obviously even with this condition that could still happen (the cancellation token could have cancellation requested _and_ there could be a funceval).

Reply:
Yes, a safe func-eval abort uses exactly the same runtime mechanism and if you have both at the same time, behavior may not be perfect.  Originally, I planned to save the stack pointer to the `ControlledExecution.Run` frame on the native thread object and unwind till that frame.  That would allow to support a hierarchy of abort requests and allow recursive `ControlledExecution.Run` calls, where, for example, an abort request for the inner `Run` could be 'upgraded' to an abort request for the outer `Run`, but not vice versa.  The feedback I received from Jan Kotas and Kevin Ransom was that we do not need to support recursive `Run` calls, so I could not justify that extra complexity, and a simultaneous fun-eval abort seems to be a very corner case given that we do not want this API to be used except for very specific cases.

Reply:
> I'd prefer TAE instances never find their way to user code, even as inner exceptions.

The code we abort may catch a TAE and wrap it.  That happens, for example, if the `CancellationTokenSource.Cancel` method is called by the `action` itself.  In that case the TAE is wrapped in an `AggregateException`.